### DISCUSSION TITLE
How to install in pyodide

### DISCUSSION THREAD

**Discussion by aszenz** [perspective-dev/perspective#2880] [Q&A] [answered]
Does perspective support pyodide distribution of python, how would I use perspective in wasm notebooks

Tried `micropip.install('perspective-python')` doesn't work

**Comment by texodus** [comment]
[Please refer to the Pyodide docs](https://pyodide.org/en/stable/usage/loading-packages.html#loading-packages) for installing emscripten/wasm32 wheels. We do not publish emscripten wheels anywhere except [GitHub Releases](https://github.com/finos/perspective/releases/tag/v3.2.0), which is not a CDN.

**Reply by aszenz** [reply]
Ok, so i tried using the wasm wheel url from github releases:
 
```python
micropip.install("https://github.com//finos/perspective/releases/download/v3.2.0/perspective_python-3.2.0-cp39-abi3-emscripten_3_1_58_wasm32.whl")
``` 
but that failed with a CORS error. 

Next I tried self hosting the wheel, which did work until I tried to create a table.

```python
import micropip
await micropip.install("/psp_wasm.whl")
import perspective
perspective.table("name,age\naszen,99")
```

> pyodide.asm.js:10 Uncaught (in promise) TypeError: getWasmTableEntry(...) is not a function

Btw what's the reason for not publishing wasm wheel

**Reply by timkpaine** [reply]
you cannot publish a wasm wheel to pypi. It sounds like you should spend a bit more time reading the pyodide documentation to see what is and is not supported. The page linked earlier has answers to all of the questions and issues you've asked so far.

**Reply by aszenz** [reply]
What would it take for pyodide to include perspective in it's pkg manager, like it does `duckdb`  which is also wasm based

**Reply by aszenz** [reply]
I had a chance to investigate this further, to install the pkg via pyodide, perspective would need to host the wasm wheel somewhere, and then create a PR at https://github.com/pyodide/pyodide/tree/main/packages for adding the package to their list.

An example pkg meta.yml file from duckdb:

https://github.com/pyodide/pyodide/blob/main/packages/duckdb/meta.yaml#L1-L15

@timkpaine / @texodus Is their any interest in doing this?

**Reply by texodus** [reply, marked as answer]
No, we're not interested in hosting this at this time. The Pyodide platform is still experimental and it is IMO too much maintenance work to keep it up to date for our OSS project, and having a half-broken or out-of-date version of Perspective floating around in Pyodide's dist is not worth the risk.

**Comment by aszenz** [comment]
For anyone that wants to use perspective in Pyodide, make sure that your version of Pyodide matches the one supported by perspective.

To load the whl over http using relative url use the prefix `http:`

```python
import micropip
whl_path = "http:/public/perspective_python-3.3.4-cp39-abi3-emscripten_3_1_58_wasm32.whl"
await micropip.install(whl_path)
import perspective
psp_table = perspective.table("name,age\naszen,99")
print(psp_table.schema())
```
Here's a sample repo demonstrating using perspective in marimo https://aszenz.github.io/perspective-pyodide-sample/
